lib/Yield/CrossRelations.cpp

```diff
@@ -126,9 +126,30 @@ yield::Graph yield::CrossRelations::toYieldGraph() const {
                                                      ranks::BasicBlock,
                                                      ranks::Instruction);
     revng_assert(FromAddress.has_value() && ToAddress.has_value());
-    LookupHelper.at(*FromAddress)->addSuccessor(LookupHelper.at(*ToAddress));
+    LookupHelper.at(*ToAddress)->addSuccessor(LookupHelper.at(*FromAddress));
   };
   conversionHelper(*this, AddNode, AddEdge);
 
+  llvm::SmallVector<yield::Graph::Node *, 4> Entries;
+  for (yield::Graph::Node *Node : Result.nodes()) {
+    // Ensure there are no nodes with invalid addresses.
+    revng_assert(Node->Address.isValid());
+
+    // Gather all the entry nodes.
+    if (!Node->hasPredecessors())
+      Entries.push_back(Node);
+  }
+
+  revng_assert(!Entries.empty());
+  if (Entries.size() > 1) {
+    // Add an artificial "root" node to make sure there's a single entry point.
+    yield::Graph::Node *Root = Result.addNode();
+    for (yield::Graph::Node *Entry : Entries)
+      Root->addSuccessor(Entry);
+    Result.setEntryNode(Root);
+  } else {
+    Result.setEntryNode(Entries.front());
+  }
```

Comment:
Why do we need to do this? Is it just temporary for layout purposes or is the entry node visible?
If we want this, we probably should have a dummy node in all situations.

Reply:
It's not visible (the flag as to whether to show "empty" nodes or not is part of the `exportGraph` helper signature), this is an optimization making using the graph with llvm's utilities a lot easier - you don't need to worry about preprocessing the graph before calling `llvm::df_begin` and stuff - you already have the perfect starting node.

Reply:
I think the current implementation does not consider loops.
In a general graph, creating an entry node of the graph is not straightforward due to the presence of loops.
In order to make sure all the nodes are reachable, you should iterate over all the SCCs (see `scc_begin`), check if an SCC has any node that jumps into it and, if not, elect a random node of a loop as the entry of the loop and connect it to the root of your graph.

If you implement this, please do it only using graph traits so we can reuse.
If this seems too much work and makes you think you don't really need this *entry node* feature, that's fine too.
In other situations, we obtained the same result by simply creating an entry node and connecting it to *all* the nodes in the graph. This has the property of not having to elect random nodes as entry nodes of loops and it's not that bad since it preserves the dominator tree of the rest of the graph.

Reply:
Or maybe this is a call tree and there are no loops?

Reply:
Actually, I don't need the "perfect" entry point - if some nodes are not reachable from here - so be it. All I care about here is being able to access all the "never called" functions from one point. I guess there is a better way to convey what's going on here.

Reply:
Uhm, if you're just interested in that, why change the graph?
You could you just make a list of those with no predecessors. Why the need for `depthFirst`?

Reply:
Yeah, I guess it wasn't a good idea to put this stuff this early in the pipeline. I might as well move it closer to the user of this "starting" node.

Reply:
Here(https://github.com/revng/revng/pull/191/commits/0f54613248446b7cae023ec816352b25e17d877b)'s a draft, hopefully that's close to what you meant initially, what do you think? @aleclearmind

Reply:
I'll close this thread since the code it's spawned around is gone. And the discussion around the entry point detection is done elsewhere.